### ISSUE TITLE
Restricted Project Naming

### ISSUE THREAD

**Issue by coopbri** [gleam-lang/gleam#1728]
## Description

When creating a new Gleam project with `gleam new`, there are situations where the `gleam` CLI will respond with errors due to disallowed project naming patterns:

<img src="https://user-images.githubusercontent.com/20056195/184559681-372e9ce0-7d7a-49c8-a650-470b8995cf8e.png" width="400px" />

Context: based on [this Twitter thread](https://twitter.com/0xcoop/status/1558609481352859648)

## Ideas

1.  Document this in the getting started guide/docs (may be a good temporary fix)
    - Mention reserved keywords situations (`gleam` prefix) as well as underscore convention
2.  Combine the error messages (though might be too much output to the user)
3.  Allow more project naming patterns; decouple Gleam project names from their parent folder name, allowing project directory names such as `gleam-demo`, `gleam`, and `gleam_demo`
    - I like this idea the most of the three proposed here; this is allowed in, for example, a new Rust/Cargo project. You can create projects with analogous name `rust-demo`, `rust`, and `rust_demo`:

      <img src="https://user-images.githubusercontent.com/20056195/184559734-97e7309e-8c17-425f-87af-494a706ee46a.png" width="500px" />

Let me know if you have other ideas. I'm happy to take this on after a consensus is reached!

**Comment by lpil** [comment]
Thanks for the report, this must have been very annoying.

I would suggest that in these instances we could print something like this

"We were not able to create your project as `gleam` as etc etc....

Would you like is to create a project named `my_gleam` in a directory named `gleam`? (Y/n)"

> decouple Gleam project names from their parent folder name, allowing project directory names such as gleam-demo, gleam, and gleam_demo

This is already the case! Unfortunately we take the project name as the argument here and use the director name as the same.

**Comment by inoas** [comment]
Maybe instead of saying what it should not be it should say what it should be:

Like:

> Project name must be **lowercase**, **alphanumeric**, may **not** start with a number and may not be named **gleam** or begin with **gleam_**.

**Comment by coopbri** [comment]
> Thanks for the report, this must have been very annoying.
> 
> I would suggest that in these instances we could print something like this
> 
> "We were not able to create your project as `gleam` as etc etc....
> 
> Would you like is to create a project named `my_gleam` in a directory named `gleam`? (Y/n)"
> 
> > decouple Gleam project names from their parent folder name, allowing project directory names such as gleam-demo, gleam, and gleam_demo
> 
> This is already the case! Unfortunately we take the project name as the argument here and use the director name as the same.

Not annoying at all! No worries.

I think coupling the solution from @inoas with the automated `(Y/n)` prompt suggested by @lpil might be a nice approach. Basically conveying all of the restrictions in one go while also automating a suggested project name.

**Comment by lpil** [comment]
Fab, I agree. Let's do it

**Comment by lpil** [comment]
The solution: make it suggest an alternative name. The full list of problems could be suggested in the text also, though the main problem would need to be presented first and foremost as today.

**Comment by lpil** [comment]
Yes please! That'd be great. Do ask any questions if you have any while working on it.

**Comment by Tar-Tarus** [comment]
Hey @lpil ,
I've been working on #3395 and this issue here seemed close enough to tackle next. 
I've already started hacking around for a solution, would it be alright to take this up?
I also have some initial questions but I guess it would be best to ask those in a PR when the code suggestion can be checked out in context as well...

**Comment by lpil** [comment]
Great! Yes please @Tar-Tarus !!

**Comment by diemogebhardt** [comment]
I had a look into it in PR https://github.com/gleam-lang/gleam/pull/3913.

**Comment by diemogebhardt** [comment]
This issue might be closed now by having merged PR https://github.com/gleam-lang/gleam/pull/3913, @lpil?

**Comment by lpil** [comment]
Thank you!!
